scripts/opentitan/verilate.py

```diff
@@ -0,0 +1,590 @@
+from ot.util.log import ColorLogFormatter, configure_loggers
+
+VeriVcpDescriptor = tuple[str, BufferedRandom, bytearray, logging.Logger]
+
+DEFAULT_TIMEOUT = 10.0 * 60
+"""Default execution timeout."""
+
+
+class VtorFileManager:
+    """File manager.
+
+       Handle temporary directories and files.
+
+       :param keep_temp: do not automatically discard generated files on exit
+       :param tmp_dir: store the temporary files in the specified directory
+    """
+
+    def __init__(self, keep_temp: bool = False, tmp_dir: Optional[str] = None):
+        self._log = getLogger('vtor.file')
+        self._keep_temp = keep_temp
+        self._base_tmp_dir = tmp_dir
+        self._tmp_dir: Optional[str] = None
+        register(self._cleanup)
+
+    @property
+    def tmp_dir(self) -> str:
+        """Provides the main temporary directory for executing Verilator."""
+        if not self._tmp_dir:
+            self._tmp_dir = mkdtemp(prefix='verilator_ot_dir_',
+                                    dir=self._base_tmp_dir)
+        return self._tmp_dir
+
+    def _cleanup(self):
+        if self._tmp_dir and isdir(self._tmp_dir):
+            if self._keep_temp:
+                self._log.warning('Preserving temp dir %s', self._tmp_dir)
+                return
+            self._log.debug('Removing temp dir %s', self._tmp_dir)
+            rmtree(self._tmp_dir)
+            self._tmp_dir = None
+
+
+class VtorExecuter:
+    """Verilator wrapper.
+
+        -c|--term-after-cycles=N
+    """
+
+    VMEM_OFFSET = 0
+    """Offset when converting BM test to VMEM file."""
+
+    ANSI_CRE = re.compile(rb'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]')
+    """ANSI escape sequences."""
+
+    DV_CRE = re.compile(r'^(\d+):\s\(../.*?\)\s(.*)$')
+    """DV macro messages."""
+
+    def __init__(self, vfm: VtorFileManager, verilator: str,
+                 profile: Optional[str], debug: bool = False):
+        self._log = getLogger('vtor.exec')
+        self._vlog = getLogger('vtor.out')
+        self._profile = profile
+        self._debug = debug
+        self._verilator = verilator
+        self._fm = vfm
+        # parsed communication ports from Verilator
+        self._ports: dict[str, Union[int, str]] = {}
+        # where Verilator stores the execution log file
+        self._xlog_path: Optional[str] = None
+        # where to copy the execution log file, if requested
+        self._dest_log_path: Optional[str] = None
+        # where to create a symbolic list to the execution log file
+        self._link_log_path: Optional[str] = None
+        self._vcps: dict[int, VeriVcpDescriptor] = {}
+        self._poller = sel_poll()
+        self._resume = False
+        self._threads: list[Thread] = []
+
+    @classmethod
+    def _simplifly_cli(cls, args: list[str]) -> str:
+        """Shorten the Verilator commmand line."""
+        return ' '.join(((','.join((basename(b) for b in a.split(',')))
+                        for a in args)))
```

Comment:
This is kind of hard to read. Maybe better as something like:

```suggestion
        args = [','.join(basename(item) for item in arg.split(','))
                for arg in args]
        return ' '.join(args)
```

Reply:
Would it be ok to read with a generator?:
````py
        arggen = (','.join((basename(b) for b in a.split(','))) for a in args)
        return ' '.join(arggen)
````

Reply:
That's fine - I think I (personally) would still prefer using some more descriptive names than `a` and `b` though just to make it slightly clearer what's going on.

Reply:
````py
arggen = (','.join((basename(path) for path in arg.split(','))) for arg in args)
return ' '.join(arggen)
````
?

Reply:
Much easier to understand for me, thanks for changing it.

Reply:
I should quit the bad habit of using one-char variable name in comprehension lists only because pylint allows it. Not a good idea, sorry about this.